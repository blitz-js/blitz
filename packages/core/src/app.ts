/*
 * IF YOU CHANGE THIS FILE
 *    You also need to update the rewrite map in
 *    packages/babel-preset/src/rewrite-imports.ts
 */
export {default as App} from "next/app"
