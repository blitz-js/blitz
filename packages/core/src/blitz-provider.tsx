import React, {FC, useRef} from "react"
import {QueryClient, QueryClientProvider} from "react-query"
import {Hydrate, HydrateOptions} from "react-query/hydration"
import {queryClient} from "./utils/react-query-utils"

export type BlitzProviderProps = {
  client?: QueryClient
  contextSharing?: boolean
  dehydratedState?: unknown
  hydrateOptions?: HydrateOptions
}

export const BlitzProvider: FC<BlitzProviderProps> = ({
  client,
  contextSharing = false,
  dehydratedState,
  hydrateOptions,
  children,
}) => {
  const queryClientRef = useRef<QueryClient>()
  if (!queryClientRef.current) {
    queryClientRef.current = queryClient
  }

  return (
    <QueryClientProvider client={client ?? queryClient} contextSharing={contextSharing}>
      <Hydrate state={dehydratedState} options={hydrateOptions}>
        {children}
      </Hydrate>
    </QueryClientProvider>
  )
}
