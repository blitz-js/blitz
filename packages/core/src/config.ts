/*
 * IF YOU CHANGE THIS FILE
 *    You also need to update the rewrite map in
 *    packages/babel-preset/src/rewrite-imports.ts
 */
export {default as getConfig, setConfig} from "next/config"
