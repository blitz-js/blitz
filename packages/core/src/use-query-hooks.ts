import Router from "next/router"
import {
  InfiniteQueryConfig as RQInfiniteQueryConfig,
  InfiniteQueryResult,
  PaginatedQueryConfig,
  PaginatedQueryResult,
  queryCache,
  QueryConfig,
  QueryResult,
  useInfiniteQuery as useInfiniteReactQuery,
  usePaginatedQuery as usePaginatedReactQuery,
  useQuery as useReactQuery,
} from "react-query"
import {FirstParam, PromiseReturnType, QueryFn} from "./types"
import {
  defaultQueryConfig,
  getQueryCacheFunctions,
  getQueryKey,
  QueryCacheFunctions,
  sanitize,
} from "./utils/react-query-utils"

Router.events.on("routeChangeComplete", async () => {
  await queryCache.invalidateQueries()
})

// -------------------------
// useQuery
// -------------------------
type RestQueryResult<TResult> = Omit<QueryResult<TResult>, "data"> & QueryCacheFunctions<TResult>

export function useQuery<T extends QueryFn, TResult = PromiseReturnType<T>>(
  queryFn: T,
  params: FirstParam<T>,
  options?: QueryConfig<TResult>,
): [TResult, RestQueryResult<TResult>] {
  if (typeof queryFn === "undefined") {
    throw new Error("useQuery is missing the first argument - it must be a query function")
  }

  const enhancedResolverRpcClient = sanitize(queryFn)
  const queryKey = getQueryKey(queryFn, params)

  const {data, ...queryRest} = useReactQuery({
    queryKey,
    queryFn: (_apiUrl: string, params: any) =>
      enhancedResolverRpcClient(params, {fromQueryHook: true, alreadySerialized: true}),
    config: {
      ...defaultQueryConfig,
      ...options,
    },
  })

  const rest = {
    ...queryRest,
    ...getQueryCacheFunctions<FirstParam<T>, TResult, T>(queryFn, params),
  }

  return [data as TResult, rest as RestQueryResult<TResult>]
}

// -------------------------
// usePaginatedQuery
// -------------------------
type RestPaginatedResult<TResult> = Omit<PaginatedQueryResult<TResult>, "resolvedData"> &
  QueryCacheFunctions<TResult>

export function usePaginatedQuery<T extends QueryFn, TResult = PromiseReturnType<T>>(
  queryFn: T,
  params: FirstParam<T>,
  options?: PaginatedQueryConfig<TResult>,
): [TResult, RestPaginatedResult<TResult>] {
  if (typeof queryFn === "undefined") {
    throw new Error("usePaginatedQuery is missing the first argument - it must be a query function")
  }

  const enhancedResolverRpcClient = sanitize(queryFn)
  const queryKey = getQueryKey(queryFn, params)

  const {resolvedData, ...queryRest} = usePaginatedReactQuery({
    queryKey,
    queryFn: (_apiUrl: string, params: any) =>
      enhancedResolverRpcClient(params, {fromQueryHook: true, alreadySerialized: true}),
    config: {
      ...defaultQueryConfig,
      ...options,
    },
  })

  const rest = {
    ...queryRest,
    ...getQueryCacheFunctions<FirstParam<T>, TResult, T>(queryFn, params),
  }

  return [resolvedData as TResult, rest as RestPaginatedResult<TResult>]
}

// -------------------------
// useInfiniteQuery
// -------------------------
type RestInfiniteResult<TResult> = Omit<InfiniteQueryResult<TResult>, "resolvedData"> &
  QueryCacheFunctions<TResult>

interface InfiniteQueryConfig<TResult, TFetchMoreResult> extends RQInfiniteQueryConfig<TResult> {
  getFetchMore?: (lastPage: TResult, allPages: TResult[]) => TFetchMoreResult
}

// TODO - Fix TFetchMoreResult not actually taking affect in apps.
// It shows as 'unknown' in the params() input argumunt, but should show as TFetchMoreResult
export function useInfiniteQuery<
  T extends QueryFn,
  TFetchMoreResult = any,
  TResult = PromiseReturnType<T>
>(
  queryFn: T,
  params: (fetchMoreResult: TFetchMoreResult) => FirstParam<T>,
  options: InfiniteQueryConfig<TResult, TFetchMoreResult>,
): [TResult[], RestInfiniteResult<TResult>] {
  if (typeof queryFn === "undefined") {
    throw new Error("useInfiniteQuery is missing the first argument - it must be a query function")
  }

  const enhancedResolverRpcClient = sanitize(queryFn)
  const queryKey = getQueryKey(queryFn, params)

  const {data, ...queryRest} = useInfiniteReactQuery({
    // we need an extra cache key for infinite loading so that the cache for
    // for this query is stored separately since the hook result is an array of results.
    // Without this cache for usePaginatedQuery and this will conflict and break.
    queryKey: [...queryKey, "infinite"],
    queryFn: (
      _apiUrl: string,
      _args: any,
      _infinite: string,
      resultOfGetFetchMore: TFetchMoreResult,
    ) => enhancedResolverRpcClient(params(resultOfGetFetchMore), {fromQueryHook: true}),
    config: {
      ...defaultQueryConfig,
      ...options,
    },
  })

  const rest = {
    ...queryRest,
    ...getQueryCacheFunctions<FirstParam<T>, TResult, T>(queryFn, params),
  }

  return [data as TResult[], rest as RestInfiniteResult<TResult>]
}
