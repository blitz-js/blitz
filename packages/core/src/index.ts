import {NextPage, NextComponentType, NextPageContext} from "next"
import {AppProps as NextAppProps} from "next/app"

export * from "./types"
export * from "./errors"
export {useQuery, usePaginatedQuery, useInfiniteQuery} from "./use-query-hooks"
export {getQueryKey, invalidateQuery} from "./utils/react-query-utils"
export {useParam, useParams} from "./use-params"
export {withRouter, RouterContext, BlitzRouter} from "./with-router"
export {useRouter} from "./use-router"
export {useRouterQuery} from "./use-router-query"
export {passportAuth} from "./passport-adapter"
export {getIsomorphicEnhancedResolver} from "./rpc"
export {useMutation} from "./use-mutation"
export {invoke, invokeWithMiddleware} from "./invoke"
export {
  getAllMiddlewareForModule,
  handleRequestWithMiddleware,
  MiddlewareResponse,
  MiddlewareRequest,
  connectMiddleware,
  Ctx,
  DefaultCtx,
} from "./middleware"
export {
  TOKEN_SEPARATOR,
  HANDLE_SEPARATOR,
  SESSION_TYPE_OPAQUE_TOKEN_SIMPLE,
  SESSION_TYPE_ANONYMOUS_JWT,
  SESSION_TOKEN_VERSION_0,
  COOKIE_ANONYMOUS_SESSION_TOKEN,
  COOKIE_SESSION_TOKEN,
  COOKIE_REFRESH_TOKEN,
  COOKIE_CSRF_TOKEN,
  COOKIE_PUBLIC_DATA_TOKEN,
  HEADER_CSRF,
  HEADER_PUBLIC_DATA_TOKEN,
  HEADER_SESSION_REVOKED,
  HEADER_CSRF_ERROR,
  LOCALSTORAGE_PREFIX,
  getAntiCSRFToken,
  useSession,
  SessionConfig, // new
  PublicData,
  SessionContext,
  DefaultPublicData,
} from "./supertokens"

// --------------------
// Exports from Next.js
// --------------------
export {
  GetStaticProps,
  GetStaticPaths,
  GetServerSideProps,
  InferGetStaticPropsType,
  InferGetServerSidePropsType,
  NextApiRequest as BlitzApiRequest,
  NextApiResponse as BlitzApiResponse,
} from "next"

export {default as Head} from "next/head"

export {default as Link} from "next/link"

export {default as Router} from "next/router"

export {
  default as Document,
  Html,
  Head as DocumentHead,
  Main,
  NextScript as BlitzScript,
  DocumentContext,
  DocumentInitialProps,
} from "next/document"

export {default as dynamic} from "next/dynamic"

export {default as ErrorComponent} from "next/error"

export type BlitzComponentType<C = NextPageContext, IP = {}, P = {}> = NextComponentType<C, IP, P>

export interface AppProps<P = {}> extends NextAppProps<P> {
  Component: BlitzComponentType<NextPageContext, any, P> & {
    getLayout?: (component: JSX.Element) => JSX.Element
  }
}
export type BlitzPage<P = {}, IP = P> = NextPage<P, IP> & {
  getLayout?: (component: JSX.Element) => JSX.Element
}
export {isLocalhost} from "./utils/index"
