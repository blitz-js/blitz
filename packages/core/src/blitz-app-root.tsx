import React, {ComponentPropsWithoutRef, useEffect} from "react"
import {useAuthorizeIf} from "./auth/auth-client"
import {publicDataStore} from "./auth/public-data-store"
import {BlitzProvider} from "./blitz-provider"
import {Head} from "./head"
import {AppProps, BlitzPage} from "./types"

const customCSS = `
  body::before {
    content: "";
    display: block;
    position: fixed;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    z-index: 99999;
    background-color: white;
  }

  .blitz-first-render-complete body::before {
    display: none;
  }
`
const noscriptCSS = `
  body::before {
    content: none
  }
`

const NoPageFlicker = () => {
  return (
    <Head>
      <style dangerouslySetInnerHTML={{__html: customCSS}} />
      <noscript>
        <style dangerouslySetInnerHTML={{__html: noscriptCSS}} />
      </noscript>
    </Head>
  )
}

export function withBlitzInnerWrapper(Page: BlitzPage) {
  const BlitzInnerRoot = (props: ComponentPropsWithoutRef<BlitzPage>) => {
    useAuthorizeIf(Page.authenticate === true)

    return <Page {...props} />
  }
  for (let [key, value] of Object.entries(Page)) {
    ;(BlitzInnerRoot as any)[key] = value
  }
  if (process.env.NODE_ENV !== "production") {
    BlitzInnerRoot.displayName = `BlitzInnerRoot`
  }
  return BlitzInnerRoot
}

export function withBlitzAppRoot(UserAppRoot: React.ComponentType<any>) {
  const BlitzOuterRoot = (props: AppProps) => {
    if (typeof window !== "undefined") {
      if (publicDataStore.getData().userId) {
        if (props.Component.redirectAuthenticatedTo) {
          window.location.replace(props.Component.redirectAuthenticatedTo)
        }
      } else {
        const authenticate = props.Component.authenticate
        if (
          authenticate &&
          typeof authenticate === "object" &&
          typeof authenticate.redirectTo === "string"
        ) {
          const url = new URL(authenticate.redirectTo, window.location.href)
          url.searchParams.append("next", window.location.pathname)
          window.location.replace(url.toString())
        }
      }
    }
    const component = React.useMemo(() => withBlitzInnerWrapper(props.Component), [props.Component])

    const noPageFlicker =
      props.Component.suppressFirstRenderFlicker ||
      props.Component.authenticate !== undefined ||
      props.Component.redirectAuthenticatedTo

    useEffect(() => {
      document.documentElement.classList.add("blitz-first-render-complete")
    }, [])

    return (
      <BlitzProvider dehydratedState={props.pageProps.dehydratedState}>
        {noPageFlicker && <NoPageFlicker />}
        <UserAppRoot {...props} Component={component} />
      </BlitzProvider>
    )
  }
  return BlitzOuterRoot
}
