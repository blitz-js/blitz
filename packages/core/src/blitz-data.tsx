import {getConfig} from "@blitzjs/config"
import htmlescape from "htmlescape"
import React from "react"
import {isClient} from "./utils"

export type BlitzRuntimeData = {
  suspenseEnabled: boolean
  sessionCookiePrefix: string
}

export function _getBlitzRuntimeData(): BlitzRuntimeData {
  const config = getConfig()
  return {
    sessionCookiePrefix: (config._meta.packageName || "blitz").replace(/[^a-zA-Z0-9-_]/g, "_"),
    suspenseEnabled: config.experimental?.reactRoot !== false,
  }
}

export function getBlitzRuntimeData() {
  if (isClient && !process.env.JEST_WORKER_ID) {
    return window.__BLITZ_DATA__
  } else {
    if (!global.__BLITZ_DATA__) {
      global.__BLITZ_DATA__ = _getBlitzRuntimeData()
    }
    return global.__BLITZ_DATA__
  }
}

// Automatically deserialize __BLITZ_DATA__ in a browser environment
if (isClient && !process.env.JEST_WORKER_ID) {
  deserializeAndSetBlitzDataOnWindow()
}

export function deserializeAndSetBlitzDataOnWindow() {
  try {
    const data: typeof window["__BLITZ_DATA__"] = JSON.parse(
      document.getElementById("__BLITZ_DATA__")!.textContent!,
    )
    window.__BLITZ_DATA__ = data
  } catch (e) {
    console.error(
      "Error deserializing __BLITZ__DATA__. Make sure you have a custom _document.js/tsx page that uses <BlitzScript/>",
      e,
    )
  }
}

export function BlitzData() {
  return (
    <script
      id="__BLITZ_DATA__"
      type="application/json"
      dangerouslySetInnerHTML={{
        __html: htmlescape(_getBlitzRuntimeData()),
      }}
    />
  )
}
