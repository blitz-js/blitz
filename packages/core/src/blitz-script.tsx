import {NextScript} from "next/document"
import React, {ComponentPropsWithoutRef} from "react"
import {BlitzData} from "./blitz-data"

export type BlitzScriptProps = ComponentPropsWithoutRef<typeof NextScript>

export function BlitzScript(props: BlitzScriptProps) {
  return (
    <>
      <BlitzData />
      <NextScript {...props} />
    </>
  )
}
