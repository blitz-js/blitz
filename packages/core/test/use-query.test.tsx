import React from "react"
import {act, render, waitForElementToBeRemoved, screen} from "./test-utils"
import {useQuery} from "../src/use-query-hooks"
import {deserialize} from "superjson"

// This enhance fn does what getIsomorphicEnhancedResolver does during build time
const enhance = (fn: any) => {
  const newFn = (...args: any) => {
    const [data, ...rest] = args
    return fn(deserialize(data), ...rest)
  }
  newFn._meta = {
    name: "testResolver",
    type: "query",
    path: "app/test",
    apiUrl: "test/url",
  }
  return newFn
}

describe("useQuery", () => {
  const setupHook = (
    params: any,
    queryFn: (...args: any) => Promise<any>,
  ): [{data?: any}, Function] => {
    let res = {}
    function TestHarness() {
      const [data] = useQuery(queryFn, params)
      Object.assign(res, {data})
      return <div id="harness">{data ? "Ready" : "Missing Dependency"}</div>
    }

    const ui = () => (
      <React.Suspense fallback="Loading...">
        <TestHarness />
      </React.Suspense>
    )

    const {rerender} = render(ui())
    return [res, () => rerender(ui())]
  }

  describe('a "query" that converts the string parameter to uppercase', () => {
    // eslint-disable-next-line require-await
    const upcase = async (args: string): Promise<string> => {
      return args.toUpperCase()
    }
    it("should work with Blitz queries", async () => {
      const [res] = setupHook("test", enhance(upcase))
      await waitForElementToBeRemoved(() => screen.getByText("Loading..."))
      await act(async () => {
        await screen.findByText("Ready")
        expect(res.data).toBe("TEST")
      })
    })

    it("shouldn't work with regular functions", () => {
      expect(() => setupHook("test", upcase)).toThrowErrorMatchingSnapshot()
    })
  })
})
