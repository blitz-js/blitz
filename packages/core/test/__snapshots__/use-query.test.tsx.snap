// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQuery a "query" that converts the string parameter to uppercase shouldn't work with regular functions 1`] = `"It looks like you are trying to use Blitz's useQuery to fetch from third-party APIs. To do that, import useQuery directly from \\"react-query\\""`;
